pilot/pkg/model/sidecar.go

```diff
@@ -0,0 +1,324 @@
+		out.allImportedDestinationRules[s.Hostname] = ps.DestinationRule(&dummyNode, s.Hostname)
+	}
+	meshGateway := map[string]bool{IstioMeshGateway: true}
+	out.allImportedVirtualServices = out.selectVirtualServices(ps.VirtualServices(&dummyNode, meshGateway))
+	return out
+}
+
+// Services returns the list of services imported across all egress listeners by this
+// Sidecar config
+func (sc *SidecarScope) Services() []*Service {
+	if sc == nil {
+		return nil
+	}
+
+	return sc.allImportedServices
+}
+
+// VirtualServices returns the list of virtual services imported by this Sidecar config
+// across all egress listeners
+func (sc *SidecarScope) VirtualServices() []Config {
+	if sc == nil {
+		return nil
+	}
+
+	return sc.allImportedVirtualServices
+}
+
+// DestinationRule returns the destination rule applicable for a given hostname
+// used by CDS code
+func (sc *SidecarScope) DestinationRule(hostname Hostname) *Config {
+	if sc == nil {
+		return nil
+	}
+
+	return sc.allImportedDestinationRules[hostname]
+}
+
+// Services returns the list of services imported by this egress listener
+func (ilw *IstioEgressListenerWrapper) Services() []*Service {
+	if ilw == nil {
+		return nil
+	}
+
+	return ilw.importedServices
+}
+
+func convertIstioListenerToWrapper(ps *PushContext, sidecarConfig *Config,
+	istioListener *networking.IstioEgressListener) *IstioEgressListenerWrapper {
+
+	out := &IstioEgressListenerWrapper{
+		IstioListener: istioListener,
+		importMap:     make(map[string]Hostname),
+	}
+
+	if istioListener.Port != nil {
+		out.ListenerPort = &Port{
+			Name:     istioListener.Port.Name,
+			Port:     int(istioListener.Port.Number),
+			Protocol: ParseProtocol(istioListener.Port.Protocol),
+		}
+	}
+
+	if istioListener.Hosts != nil {
+		for _, h := range istioListener.Hosts {
+			parts := strings.SplitN(h, "/", 2)
+			out.importMap[parts[0]] = Hostname(parts[1])
+		}
+	}
+
+	dummyNode := Proxy{
+		ConfigNamespace: sidecarConfig.Namespace,
+	}
+
+	out.importedServices = out.selectServices(ps.Services(&dummyNode))
+
+	return out
+}
+
+// GetEgressListenerForPort returns the egress listener corresponding to
+// the listener port or the catch all listener
+func (sc *SidecarScope) getEgressListenerForPort(port int) *IstioEgressListenerWrapper {
+	if sc == nil {
+		return nil
+	}
+
+	for _, e := range sc.EgressListeners {
+		if e.ListenerPort == nil || e.ListenerPort.Port == port {
+			return e
+		}
+	}
+	return nil
+}
+
+// selectVirtualServices returns the list of virtual services selected
+// across all egress listeners
+func (sc *SidecarScope) selectVirtualServices(configs []Config) []Config {
+
+	importedConfigs := make([]Config, 0)
+	for _, c := range configs {
+		configNamespace := c.Namespace
+		rule := c.Spec.(*networking.VirtualService)
+		for _, ilw := range sc.EgressListeners {
+			// Check if there is an explicit import of form ns/* or ns/host
+			if hostMatch, nsFound := ilw.importMap[configNamespace]; nsFound {
```

Comment:
ImportMap name is very confusing, and I have no idea what this code does. Far more comments. 

Why do we need a hostMatch here, to support wildcarded domains in VirtualServices ?

Reply:
this is from the sidecar API: https://github.com/istio/api/blob/release-1.1/networking/v1alpha3/sidecar.proto#L147